ctx/ctx.go

```diff
@@ -0,0 +1,143 @@
+/*
+Current request context data.
+
+`kong.ctx.shared`:
+A table that has the lifetime of the current request and is shared between
+all plugins. It can be used to share data between several plugins in a given
+request.
+
+Since only relevant in the context of a request, this table cannot be
+accessed from the top-level chunk of Lua modules. Instead, it can only be
+accessed in request phases, which are represented by the `rewrite`,
+`access`, `header_filter`, `body_filter`, `log`, and `preread` phases of
+the plugin interfaces. Accessing this table in those functions (and their
+callees) is fine.
+
+Values inserted in this table by a plugin will be visible by all other
+plugins.  One must use caution when interacting with its values, as a naming
+conflict could result in the overwrite of data.
+
+Usage:
+-- Two plugins A and B, and if plugin A has a higher priority than B's
+-- (it executes before B):
+--
+-- plugin A handler.lua
+function plugin_a_handler:access(conf)
+  kong.ctx.shared.foo = "hello world"
+--
+  kong.ctx.shared.tab = {
+    bar = "baz"
+  }
+end
+
+-- plugin B handler.lua
+function plugin_b_handler:access(conf)
+  kong.log(kong.ctx.shared.foo) -- "hello world"
+  kong.log(kong.ctx.shared.tab.bar) -- "baz"
+end
+
+
+`kong.ctx.plugin`:
+A table that has the lifetime of the current request - Unlike
+`kong.ctx.shared`, this table is **not** shared between plugins. Instead,
+it is only visible for the current plugin.
```

Comment:
since it's local to the plugin, it'd be better to do this natively in Go, to avoid roundtrips to Kong (that's why I only implemented the xxSharedXx functions).

Reply:
Duh! Very good point 😅